pypose/module/icp.py

```diff
@@ -0,0 +1,97 @@
+
+class ICP(torch.nn.Module):
+    r'''
+    Batched Iterative Closest Point (ICP) algorithm to find a rigid transformation
+    between two sets of points using Singular Value Decomposition (SVD).
+
+    Args:
+        steps (``int``, optional): The maximum number of ICP iteration steps. Default: 200.
+        tol (``double``, optional): The tolerance of the relative error used to terminate
+            the algorithm. Default: 1e-6.
+        init (``LieTensor``, optional): The initial transformation :math:`T_{init}` in
+            ``SE3type``. Default: None.
+
+    The algorithm takes two input point clouds: source point cloud (psrc) and target
+    point cloud (ptgt). The objective is to find the optimal similarity transformation
+    ( :math:`T` ) to minimize the error between the transformed source point cloud and the
+    target point cloud as shown in the equation:
+
+    .. math::
+        \begin{align*}
+            \underset{T}{\operatorname{arg\,min}} \sum_i \| p_{\mathrm{target, j}} -
+            T \cdot p_{\mathrm{source, i}}\|,
+        \end{align*}
+
+    where :math:`p_{\mathrm{source, i}}` is the ith point in the source point cloud, and
+    :math:`p_{\mathrm{target, j}}` is the cloest point to :math:`p_{\mathrm{source, i}}`
+    in the target point cloud with index j. The algorithm consists of the following steps:
+
+    1. For each point in psrc, the nearest neighbor algorithm (knn) is used to select its
+    closest point in ptgt to form the matched point pairs.
+
+    2. Singular value decomposition (SVD) algorithm is used to compute the rotation
+    and translation matrices from the matched point pairs.
+
+    3. The source point cloud (psrc) is updated using the obtained rotation and
+    translation matrices. The distance between the updated psrc and ptgt is calculated.
+
+    4. The algorithm continues to iterate through these steps until the change in the
+    calculated distance falls below the specified tolerance level or the maximum number
+    of iteration steps is reached.
+
+    Example:
+        >>> import torch, pypose as pp
+        >>> pc1 = torch.tensor([[[0., 0., 0.],
+        ...                      [1., 0., 0.],
+        ...                      [2., 0, 0.]]])
+        >>> pc2 = torch.tensor([[[0.2, 0.1, 0.],
+        ...                      [1.1397, 0.442, 0.],
+        ...                      [2.0794, 0.7840, 0.]]])
+        >>> icp = pp.module.ICP()
+        >>> icp(pc1, pc2)
+        SE3Type LieTensor:
+        LieTensor([[0.2000, 0.1000, 0.0000, 0.0000, 0.0000, 0.1736, 0.9848]])
+
+    Warning:
+        It's important to note that the solution found is only a local optimum.
+
+    '''
+    def __init__(self, steps=200, tol=1e-6, init=None):
+        super().__init__()
+        self.steps = steps
+        self.tol = tol
+        self.init = init
+        if init != None:
+            assert is_SE3(init), "The input initial transformation is not SE3Type."
+
+    def forward(self, psrc, ptgt):
```

Comment:
suggest to change args name to `source` and `target`, as the class name already says it is for point cloud.

Reply:
suggest add optional args for `ord`, `dim` to pass to `knn`.

Reply:
> suggest add optional args for `ord`, `dim` to pass to `knn`.

Yes, we can pass the 'ord' and 'dim' arguments to the KNN function, but I am slightly concerned that this may confuse users.  Retaining these arguments in the KNN function appears to be a reasonable decision for generalization purposes, as other modules or functions might necessitate greater flexibility. However, specifically for the ICP or PNP modules, since the point cloud is calculated using the SE3 Lie algebra and the point cloud format is defined as (..., N, 3), the 'dim' value must be -1 in this case. Additionally, 'ord' is naturally set to 2, as ICP involves 3D spatial geometry. these arguments are optional and should not cause significant concern at this stage. We can include them for now and, if required, make refinements in the future.

Reply:
> how about we also change the point cloud args of `knn` and `svdtf` to `source` and `target` as well?

Thanks for suggestion. I changed the variable names for svdtf, as it has the flow direction (find tf from one to the other). For knn, i'm slightly concerned that `source` and `target` may confused users, I would suggest to keep pc1 and pc2.